persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1738,6 +1751,178 @@ repsertManySql ent n = putManySql' conflictColumns fields ent n
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+--
+-- -- @since 2.12.0.1
+upsertManyWhere ::
+  forall record backend m.
+  ( backend ~ PersistEntityBackend record,
+    BackendCompatible SqlBackend backend,
+    PersistEntityBackend record ~ SqlBackend,
+    PersistEntity record,
+    MonadIO m
+  ) =>
+  -- | A list of the records you want to insert, or update
+  [record] ->
+  -- | A list of the fields you want to copy over.
+  [HandleUpdateCollision record] ->
+  -- | A list of the updates to apply that aren't dependent on the record being inserted.
+  [Update record] ->
+  -- a filter condition
+  [Filter record] ->
+  ReaderT backend m ()
+upsertManyWhere [] _ _ _ = return ()
+upsertManyWhere records fieldValues updates filters = do
+  conn <- asks projectBackend
+  uncurry rawExecute $
+    mkBulkUpsertQuery records conn fieldValues updates filters
+
+-- | This creates the query for 'upsertManyWhere'. If you
+-- provide an empty list of updates to perform, then it will generate
+-- a dummy/no-op update using the first field of the record. This avoids
+-- duplicate key exceptions.
+mkBulkUpsertQuery
+    :: (PersistEntity record, PersistEntityBackend record ~ SqlBackend)
+    => [record] -- ^ A list of the records you want to insert, or update
+    -> SqlBackend
+    -> [HandleUpdateCollision record] -- ^ A list of the fields you want to copy over.
+    -> [Update record] -- ^ A list of the updates to apply that aren't dependent on the record being inserted.
+    -> [Filter record]
+    -> (Text, [PersistValue])
+mkBulkUpsertQuery records conn fieldValues updates filters =
+  (q, recordValues <> updsValues <> copyUnlessValues <> whereVals)
+  where
+    mfieldDef x = case x of
+        CopyField rec -> Right (fieldDbToText (persistFieldDef rec))
+        CopyUnlessEq rec val -> Left (fieldDbToText (persistFieldDef rec), toPersistValue val)
+    (fieldsToMaybeCopy, updateFieldNames) = partitionEithers $ map mfieldDef fieldValues
+    fieldDbToText = escapeF . fieldDB
+    entityDef' = entityDef records
+    firstField = case entityFieldNames of
+        [] -> error "The entity you're trying to insert does not have any fields."
+        (field:_) -> field
+    entityFieldNames = map fieldDbToText (entityFields entityDef')
+    nameOfTable = escapeE . entityDB $ entityDef'
+    copyUnlessValues = map snd fieldsToMaybeCopy
+    recordValues = concatMap (map toPersistValue . toPersistFields) records
+    recordPlaceholders = Util.commaSeparated $ map (Util.parenWrapped . Util.commaSeparated . map (const "?") . toPersistFields) records
+    mkCondFieldSet n _ =
+      T.concat
+        [ n
+            ,"=COALESCE("
+              ,"NULLIF("
+                ,"EXCLUDED."
+                  ,n
+                ,","
+              ,"?"
+            ,")"
+          ,")"
```

Comment:
I think the problem is here. This text will expand out to:

```
n = COALESCE( NULLIF( EXCLUDED.n, ?))
```

But what we want is:

```
n = COALESCE( NULLIF( EXCLUDED.n, ?), n)
```

```suggestion
          , ", "
          , n
          ,")"
```

Reply:
hmmm I tested this and it still didn't completely work; now I have this:
```
INSERT INTO "item"("name", "description", "price", "quantity") VALUES (?, ?, ?, ?), (?, ?, ?, ?) ON CONFLICT ("name") DO UPDATE SET "price"=EXCLUDED."price", "quantity"=COALESCE(NULLIF(EXCLUDED."quantity",?), EXCLUDED."quantity")
[PersistText "item1",PersistText "",PersistDouble 6.0,PersistInt64 0,PersistText "item2",PersistText "hello world",PersistNull,PersistInt64 0,PersistInt64 0]
```

and the test still fails with
```
expected: [Item {itemName = "item1", itemDescription = "", itemPrice = Just 6.0, itemQuantity = Nothing},Item {itemName = "item2", itemDescription = "hello world", itemPrice = Nothing, itemQuantity = Just 2}]
        but got: [Item {itemName = "item1", itemDescription = "", itemPrice = Just 6.0, itemQuantity = Just 0},Item {itemName = "item2", itemDescription = "hello world", itemPrice = Nothing, itemQuantity = Just 0}]
```

Reply:
It's so odd to me that it expects `Nothing` but it returns a `Just 0` instead and somehow that's a different thing.  When I change that parameter to `Nothing` it still fails, just in a different way

```
  1) UpsertWhere.upsertManyWhere only copies passing values
       expected: [Item {itemName = "item1", itemDescription = "", itemPrice = Just 6.0, itemQuantity = Nothing},Item {itemName = "item2", itemDescription = "hello world", itemPrice = Nothing, itemQuantity = Just 2}]
        but got: [Item {itemName = "item1", itemDescription = "", itemPrice = Just 6.0, itemQuantity = Nothing},Item {itemName = "item2", itemDescription = "hello world", itemPrice = Nothing, itemQuantity = Nothing}]
```